File: src/src/Systems/Voter.cs
using System;
using Models;
using Godot;

namespace Systems
{
    public class Voter
    {
        public float Loyalty { get; set; } = 50f; // Influence by player/actions
        public float Independence { get; set; } = 50f; // Tendency to follow stats vs. player influence

        public bool Supports(Candidate candidate, Region region)
        {
            // Calculate support based on candidate platform and region stats
            float supportScore = 0f;
            foreach (var platform in candidate.Platform)
            {
                if (region.Stats.GetStatValue(platform.Key) >= platform.Value)
                {
                    supportScore += 10f; // Higher support if region stat meets/exceeds candidate platform
                }
                else
                {
                    supportScore -= 5f; // Lower support otherwise
                }
            }

            // Incorporate loyalty and independence
            supportScore += (Loyalty / 100f) * 10f;
            supportScore -= (Independence / 100f) * 5f;

            // Determine support based on supportScore threshold
            return supportScore > 0f;
        }
    }
}

File: src/Systems/Voter.cpp
#include "Voter.h"

#include <Godot.hpp>
#include "Region.h"

namespace Systems {

using namespace godot;

void Voter::_register_methods() {
    register_method("supports", &Voter::supports);
}

Voter::Voter() {}
Voter::~Voter() {}

void Voter::_init() {
    // Initialization if needed
}

void Voter::set_loyalty(float value) {
    loyalty = Math::clamp(value, 0.0f, 100.0f);
}

float Voter::get_loyalty() const {
    return loyalty;
}

void Voter::set_independence(float value) {
    independence = Math::clamp(value, 0.0f, 100.0f);
}

float Voter::get_independence() const {
    return independence;
}

bool Voter::supports(Candidate *candidate, Region *region) {
    float support_score = 0.0f;
    godot::Dictionary platform = candidate->get_platform();

    godot::Array keys = platform.keys();
    for(int i = 0; i < keys.size(); i++) {
        godot::String key = keys[i];
        float value = platform[key];
        float region_stat = region->get_stats()->get_stat_value(key);
        if(region_stat >= value) {
            support_score += 10.0f;
        } else {
            support_score -= 5.0f;
        }
    }

    support_score += (loyalty / 100.0f) * 10.0f;
    support_score -= (independence / 100.0f) * 5.0f;

    return support_score > 0.0f;
}

} // namespace Systems 